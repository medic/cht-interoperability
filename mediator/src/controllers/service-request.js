const axios = require('axios');
const {FHIR} = require('../../config');
const {url: fhirUrl, password: fhirPassword, username: fhirUsername} = FHIR;
const logger = require('../../logger');

async function createServiceRequest(request) {
  try {
    const {patientId} = request;

    const options = {
      auth: {
        username: fhirUsername,
        password: fhirPassword,
      }
    };

    const res = await axios.get(`${fhirUrl}/Patient/${patientId}`, options);

    if (res.status !== 200) {
      return {status: res.status, data: res.data};
    }

    // todo => @njogz to add task creation process.

    return {status: res.status, data: res.data};
  } catch ({response: res}) {
    logger.error(JSON.stringify(res.data, null, 4));
    return {status: res.status, data: res.data};
  }
}

module.exports = {
  createServiceRequest,
};
