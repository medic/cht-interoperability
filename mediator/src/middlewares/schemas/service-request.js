const joi = require('joi');

const createServiceSchema = joi.object({
  patientId: joi.number().required()
});

module.exports = {
  createServiceSchema,
};
