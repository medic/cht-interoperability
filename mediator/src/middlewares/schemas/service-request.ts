import joi from 'joi';

const RequesterSchema = joi.object({
  reference: joi.string().regex(/Organization\/\S+/),
});

export const ServiceRequestSchema = joi.object({
  intent: joi.string().required(),
  subject: joi
    .object({
      reference: joi.string().regex(/Patient\/\S+/),
    })
    .required(),
  requester: RequesterSchema.required(),
});
