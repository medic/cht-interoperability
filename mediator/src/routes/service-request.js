const {Router} = require('express');
const {createServiceRequest} = require('../controllers/service-request');
const {requestHandler} = require('../utils/request');
const {validateBodyAgainst} = require('../middlewares');
const {createServiceSchema} = require('../middlewares/schemas/service-request');

const router = Router();

router.post(
  '/',
  validateBodyAgainst(createServiceSchema),
  requestHandler(req => createServiceRequest(req.body))
);

module.exports = router;
