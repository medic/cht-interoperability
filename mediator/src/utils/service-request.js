const url = require('url');

function requestHandler(handler) {
  return  (req, res) => {
    handler(req)
    .then(({status, data}) => {
      res.status(status).send(data);
    });
  };
}

function generateApiUrl(chtUrl, username, password) {
  const parsedUrl = url.parse(chtUrl);
  const {protocol, hostname, port} = parsedUrl;
  const apiURL = `${protocol}//${username}:${password}@${hostname}/api/v2/records`;

  return apiURL;
}

module.exports = {
  requestHandler,
  generateApiUrl,
};
