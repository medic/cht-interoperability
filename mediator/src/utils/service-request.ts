import { Request, Response } from 'express';
import url from 'url';

type RequestHandler = (req: Request) => Promise<{status: number, data: any}>;

export function requestHandler(handler: RequestHandler) {
  return  (req: Request, res: Response) => {
    handler(req)
    .then(({status, data}) => {
      res.status(status).send(data);
    });
  };
}

export function generateApiUrl(chtUrl: string, username: string, password: string) {
  const parsedUrl = url.parse(chtUrl);
  const {protocol, hostname, port} = parsedUrl;
  const apiURL = `${protocol}//${username}:${password}@${hostname}/api/v2/records`;

  return apiURL;
}
