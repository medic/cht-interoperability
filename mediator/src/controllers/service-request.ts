import axios from "axios";
import { logger } from "../../logger";
import { FHIR, CHT } from "../../config";
import { generateFHIRSubscriptionResource } from "../utils/subscription";
import { generateApiUrl } from "../utils/service-request";
import https from "https";

const { url: fhirUrl, password: fhirPassword, username: fhirUsername } = FHIR;

type ServiceRequest = {
  patient_id: string;
  callback_url: string;
};

async function createServiceRequest(request: ServiceRequest) {
  try {
    const { patient_id: patientId, callback_url } = request;

    let options: Record<string, any> = {
      auth: {
        username: fhirUsername,
        password: fhirPassword,
      },
    };

    const res = await axios.get(`${fhirUrl}/Patient/?identifier=${patientId}`, options);

    if (res.status !== 200) {
      return { status: res.status, data: res.data };
    }

    // generate subscription resource
    const FHIRSubscriptionResource = generateFHIRSubscriptionResource(
      patientId,
      callback_url
    );
    const subscriptionRes = await axios.post(
      `${fhirUrl}/Subscription`,
      FHIRSubscriptionResource,
      options
    );

    if (subscriptionRes.status !== 201) {
      return { status: subscriptionRes.status, data: subscriptionRes.data };
    }

    // call the CHT API to set up the follow up task
    const chtApiUrl = generateApiUrl(CHT.url, CHT.username, CHT.password);
    options = {
      httpsAgent: new https.Agent({
        rejectUnauthorized: false,
      }),
    };
    
    const body = {
      _meta: {
        form: "interop_follow_up",
      },
      patient_uuid: patientId,
    };
    
    const chtRes = await axios.post(chtApiUrl, body, options);

    if (chtRes.data.success !== true) {
      // TODO: delete the subscription
      return { status: 500, data: { message: "unable to create the follow up task"} };
    }

    logger.info(JSON.stringify(chtRes.data, null, 4));

    return { status: res.status };
  } catch (err) {
    logger.error(`Error: ${err}`);
    return { status: 500 };
  }
}

module.exports = {
  createServiceRequest,
};
