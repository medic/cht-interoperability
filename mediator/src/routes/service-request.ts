import { validateBodyAgainst } from '../middlewares';
import { requestHandler } from '../utils/request';

const { Router } = require('express');
const { createServiceRequest } = require('../controllers/service-request');
const {
  createServiceSchema,
} = require('../middlewares/schemas/service-request');

const router = Router();

router.post(
  '/',
  validateBodyAgainst(createServiceSchema),
  requestHandler((req) => createServiceRequest(req.body))
);

export default router;
