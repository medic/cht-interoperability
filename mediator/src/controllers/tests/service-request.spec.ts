import {
  mockGetFHIRPatientResource,
  mockGetFHIROrgEndpointResource,
  mockCreateFHIRSubscriptionResource,
  mockCreateChtRecord,
  mockDeleteFhirSubscription,
} from './utils';
import { createServiceRequest } from '../service-request';
import { ServiceRequestFactory } from '../../middlewares/schemas/tests/fhir-resource-factories';

jest.mock('axios');
jest.mock('../../../logger');

describe('ServiceRequest controllers', () => {
  const request: fhir4.ServiceRequest = ServiceRequestFactory.build();

  describe('createServiceRequest', () => {
    it('creates a new subscriptions resource on fhir and cht record', async () => {
      const recordRes = { status: 201, data: { success: true } };
      const patientRes = { status: 201, data: { resourceType: 'Patient' } };
      const endpointRes = { status: 201, data: { resourceType: 'Endpoint' } };
      const subscriptionRes = {
        status: 201,
        data: { resourceType: 'Subscription' },
      };

      mockGetFHIRPatientResource.mockResolvedValueOnce(patientRes as any);
      mockGetFHIROrgEndpointResource.mockResolvedValueOnce(endpointRes as any);
      mockCreateFHIRSubscriptionResource.mockResolvedValueOnce(
        subscriptionRes as any
      );
      mockCreateChtRecord.mockResolvedValueOnce(recordRes as any);
      mockDeleteFhirSubscription.mockResolvedValueOnce({} as any);

      const res = await createServiceRequest(request);

      expect(res.status).toBe(subscriptionRes.status);
      expect(res.data).toBe(subscriptionRes.data);
    });

    it('returns default status and data when not provided in error', async () => {
      const patientRes = { message: 'message' };

      mockGetFHIRPatientResource.mockRejectedValueOnce(patientRes);

      const res = await createServiceRequest(request);

      expect(res.status).toBe(500);
      expect(res.data.message).toBe(patientRes.message);
    });


    it('returns status and data of any axios error', async () => {
      const patientRes = { status: 200, data: { resourceType: 'Patient' } };

      mockGetFHIRPatientResource.mockRejectedValueOnce(patientRes);

      const res = await createServiceRequest(request);

      expect(res.status).toBe(patientRes.status);
      expect(res.data).toBe(patientRes.data);
    });

    it('deletes fhir subscription resource if cht record creation fails', async () => {
      const recordRes = { status: 200, data: { success: false } };
      const patientRes = { status: 200, data: { resourceType: 'Patient' } };
      const endpointRes = { status: 200, data: { resourceType: 'Endpoint' } };
      const subscriptionRes = {
        status: 201,
        data: { resourceType: 'Subscription' },
      };

      mockGetFHIRPatientResource.mockResolvedValueOnce(patientRes as any);
      mockGetFHIROrgEndpointResource.mockResolvedValueOnce(endpointRes as any);
      mockCreateFHIRSubscriptionResource.mockResolvedValueOnce(
        subscriptionRes as any
      );
      mockCreateChtRecord.mockResolvedValueOnce(recordRes as any);
      mockDeleteFhirSubscription.mockResolvedValueOnce({} as any);

      const res = await createServiceRequest(request);

      expect(res.status).toBe(500);
      expect(res.data.message).toMatchInlineSnapshot(
        `"Unable to create the follow up task"`
      );
    });
  });
});
