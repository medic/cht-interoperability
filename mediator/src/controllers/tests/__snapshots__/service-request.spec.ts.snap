// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createServiceRequest creates a service request when given valid request document 1`] = `
[MockFunction] {
  "calls": [
    [
      "http://openhim-core:5001/fhir/Patient/?identifier=PATIENT_ID",
      {
        "auth": {
          "password": "interop-password",
          "username": "interop-client",
        },
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`createServiceRequest creates a service request when given valid request document 2`] = `
[MockFunction] {
  "calls": [
    [
      "http://openhim-core:5001/fhir/Subscription",
      {
        "channel": {
          "endpoint": "CALLBACK_URL",
          "header": [
            "Content-Type: application/fhir+json",
          ],
          "payload": "application/fhir+json",
          "type": "rest-hook",
        },
        "criteria": "Encounter?identifier=PATIENT_ID",
        "id": "PATIENT_ID",
        "reason": "Follow up request for patient",
        "resourceType": "Subscription",
        "status": "requested",
      },
      {
        "auth": {
          "password": "interop-password",
          "username": "interop-client",
        },
      },
    ],
    [
      "http:/admin:password@nginx/api/v2/records",
      {
        "_meta": {
          "form": "interop_follow_up",
        },
        "patient_uuid": "PATIENT_ID",
      },
      {
        "httpsAgent": Agent {
          "_events": {
            "free": [Function],
            "newListener": [Function],
          },
          "_eventsCount": 2,
          "_maxListeners": undefined,
          "_sessionCache": {
            "list": [],
            "map": {},
          },
          "defaultPort": 443,
          "freeSockets": {},
          "keepAlive": false,
          "keepAliveMsecs": 1000,
          "maxCachedSessions": 100,
          "maxFreeSockets": 256,
          "maxSockets": Infinity,
          "maxTotalSockets": Infinity,
          "options": {
            "path": null,
            "rejectUnauthorized": false,
          },
          "protocol": "https:",
          "requests": {},
          "scheduling": "lifo",
          "sockets": {},
          "totalSocketCount": 0,
          Symbol(kCapture): false,
        },
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`createServiceRequest returns a 500 status code if record creation on cht fails 1`] = `"unable to create the follow up task"`;

exports[`createServiceRequest returns a 500 status code if record creation on cht fails 2`] = `
[MockFunction] {
  "calls": [
    [
      "http://openhim-core:5001/fhir/Patient/?identifier=PATIENT_ID",
      {
        "auth": {
          "password": "interop-password",
          "username": "interop-client",
        },
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`createServiceRequest returns a 500 status code if record creation on cht fails 3`] = `
[MockFunction] {
  "calls": [
    [
      "http://openhim-core:5001/fhir/Subscription",
      {
        "channel": {
          "endpoint": "CALLBACK_URL",
          "header": [
            "Content-Type: application/fhir+json",
          ],
          "payload": "application/fhir+json",
          "type": "rest-hook",
        },
        "criteria": "Encounter?identifier=PATIENT_ID",
        "id": "PATIENT_ID",
        "reason": "Follow up request for patient",
        "resourceType": "Subscription",
        "status": "requested",
      },
      {
        "auth": {
          "password": "interop-password",
          "username": "interop-client",
        },
      },
    ],
    [
      "http:/admin:password@nginx/api/v2/records",
      {
        "_meta": {
          "form": "interop_follow_up",
        },
        "patient_uuid": "PATIENT_ID",
      },
      {
        "httpsAgent": Agent {
          "_events": {
            "free": [Function],
            "newListener": [Function],
          },
          "_eventsCount": 2,
          "_maxListeners": undefined,
          "_sessionCache": {
            "list": [],
            "map": {},
          },
          "defaultPort": 443,
          "freeSockets": {},
          "keepAlive": false,
          "keepAliveMsecs": 1000,
          "maxCachedSessions": 100,
          "maxFreeSockets": 256,
          "maxSockets": Infinity,
          "maxTotalSockets": Infinity,
          "options": {
            "path": null,
            "rejectUnauthorized": false,
          },
          "protocol": "https:",
          "requests": {},
          "scheduling": "lifo",
          "sockets": {},
          "totalSocketCount": 0,
          Symbol(kCapture): false,
        },
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`createServiceRequest returns status code and error message when give an invalid service request 1`] = `
{
  "message": "Invalid patient id was expecting type of 'string' or 'number' but received 'undefined'",
}
`;

exports[`createServiceRequest returns status code and error when server fails to create a subscription 1`] = `
{
  "message": undefined,
}
`;

exports[`createServiceRequest returns status code and error when server fails to create a subscription 2`] = `
[MockFunction] {
  "calls": [
    [
      "http://openhim-core:5001/fhir/Patient/?identifier=PATIENT_ID",
      {
        "auth": {
          "password": "interop-password",
          "username": "interop-client",
        },
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`createServiceRequest returns status code and error when server fails to create a subscription 3`] = `
[MockFunction] {
  "calls": [
    [
      "http://openhim-core:5001/fhir/Subscription",
      {
        "channel": {
          "endpoint": "CALLBACK_URL",
          "header": [
            "Content-Type: application/fhir+json",
          ],
          "payload": "application/fhir+json",
          "type": "rest-hook",
        },
        "criteria": "Encounter?identifier=PATIENT_ID",
        "id": "PATIENT_ID",
        "reason": "Follow up request for patient",
        "resourceType": "Subscription",
        "status": "requested",
      },
      {
        "auth": {
          "password": "interop-password",
          "username": "interop-client",
        },
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`createServiceRequest returns the status code of the failed susbscription creation request 1`] = `
[MockFunction] {
  "calls": [
    [
      "http://openhim-core:5001/fhir/Patient/?identifier=PATIENT_ID",
      {
        "auth": {
          "password": "interop-password",
          "username": "interop-client",
        },
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`createServiceRequest returns the status code of the failed susbscription creation request 2`] = `
[MockFunction] {
  "calls": [
    [
      "http://openhim-core:5001/fhir/Subscription",
      {
        "channel": {
          "endpoint": "CALLBACK_URL",
          "header": [
            "Content-Type: application/fhir+json",
          ],
          "payload": "application/fhir+json",
          "type": "rest-hook",
        },
        "criteria": "Encounter?identifier=PATIENT_ID",
        "id": "PATIENT_ID",
        "reason": "Follow up request for patient",
        "resourceType": "Subscription",
        "status": "requested",
      },
      {
        "auth": {
          "password": "interop-password",
          "username": "interop-client",
        },
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;
